src/icd/icdbuilder.cpp

```diff
@@ -0,0 +1,170 @@
+**
+** Copyright (C) 2018-2019 N7 Space sp. z o. o.
+** Contact: http://n7space.com
+**
+** This file is part of ASN.1/ACN Plugin for QtCreator.
+**
+** Plugin was developed under a programme and funded by
+** European Space Agency.
+**
+** This Plugin is free software: you can redistribute it and/or modify
+** it under the terms of the GNU General Public License as published by
+** the Free Software Foundation, either version 3 of the License, or
+** (at your option) any later version.
+**
+** This Plugin is distributed in the hope that it will be useful,
+** but WITHOUT ANY WARRANTY; without even the implied warranty of
+** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+** GNU General Public License for more details.
+**
+** You should have received a copy of the GNU General Public License
+** along with this program.  If not, see <http://www.gnu.org/licenses/>.
+**
+****************************************************************************/
+
+#include "icdbuilder.h"
+
+#include <QDesktopServices>
+#include <QMessageBox>
+#include <QObject>
+#include <QStringList>
+
+#include <coreplugin/icore.h>
+
+#include <projectexplorer/buildconfiguration.h>
+#include <projectexplorer/buildmanager.h>
+#include <projectexplorer/buildsteplist.h>
+#include <projectexplorer/makestep.h>
+#include <projectexplorer/project.h>
+#include <projectexplorer/projectexplorerconstants.h>
+#include <projectexplorer/target.h>
+
+#include <cmakeprojectmanager/cmakebuildstep.h>
+
+#include <asn1acn.h>
+
+using namespace Asn1Acn::Internal::Icd;
+using namespace Asn1Acn::Internal;
+using namespace ProjectExplorer;
+
+namespace {
+
+const QString GENERATED_ICD_PREFIX = "Generated ICD: ";
+const QString CMAKE_TARGETS_KEY = "CMakeProjectManager.MakeStep.BuildTargets";
+const QString ICD_TARGET_NAME = "icdFromAsn1";
+
+void openGeneratedIcd(const QString &line)
+{
+    const auto icdHtml = line.mid(GENERATED_ICD_PREFIX.length()).trimmed();
+    QDesktopServices::openUrl(QUrl::fromLocalFile(icdHtml));
+}
+
+bool icdGenerationFinished(const QString &line)
+{
+    return line.startsWith(GENERATED_ICD_PREFIX);
+}
+
+BuildStep *modifiedMakeBuildStep(BuildStepList *steps)
+{
+    auto step = steps->firstOfType<MakeStep>();
+    if (!step)
+        return nullptr;
+    step->setBuildTarget(ICD_TARGET_NAME, true);
+    return step;
+}
+
+BuildStep *modifiedCMakeBuildStep(BuildStepList *steps)
+{
+    for (auto step : steps->steps()) {
+        auto map = step->toMap();
+        if (map.contains(CMAKE_TARGETS_KEY)) {
+            map[CMAKE_TARGETS_KEY] = ICD_TARGET_NAME;
+            step->fromMap(map);
+            return step;
+        }
+    }
+    return nullptr;
+}
+
+BuildStep *modifiedBuildStep(BuildStepList *steps)
+{
+    auto makeStep = modifiedMakeBuildStep(steps);
+    if (makeStep != nullptr)
+        return makeStep;
+    return modifiedCMakeBuildStep(steps);
+}
+
+std::shared_ptr<BuildConfiguration> prepareClone(BuildConfiguration *buildConfig)
```

Comment:
Please change name to something more descriptive (maybe createSuicidalClone :P)

Reply:
createAndConnectASuicidalClone ^_^

Reply:
`wrapForProperDeletion(factory->clone(target, buildConfig));`